File: kotlin/src/main/kotlin/com/cyaim/im/client/ImCursorStore.kt
package com.cyaim.im.client

import kotlinx.serialization.Serializable
import java.io.File
import java.util.concurrent.atomic.AtomicReference

/**
 * Who a set of cursors belongs to: `(endpoint host, appId, userId)`.
 *
 * Not the device id — the store is already local to the device. [userId] is the part that matters:
 * without it, two accounts on one handset share one set of cursors, and the second one to sign in
 * silently inherits the first one's position and skips whatever the first had already consumed.
 * That is one person's cursors replaying into another person's client. CONTRACT.md §5.3.
 *
 * 作用域键必须含 userId：同一台设备换账号时，否则第二个账号会继承第一个账号的游标。
 */
public data class ImCursorScope(
    public val host: String,
    public val appId: String,
    public val userId: String,
) {
    /**
     * The identity stamped into [ImCursorSnapshot.scope] and compared on load. Spelled the same way
     * in all five SDKs, so a snapshot written by one is legible to the others.
     */
    public val key: String get() = "$host|$appId|${userId.ifBlank { "*" }}"

    /** The same identity, safe as a file name on every platform. */
    public val storageKey: String get() = key.replace(Regex("[^A-Za-z0-9._-]"), "_")

    override fun toString(): String = key

    public companion object {
        /**
         * The usual construction: the endpoint the client connects to, reduced to its host.
         *
         * The host and not the whole URL, because a deployment that moves between `wss://h/im` and
         * `wss://h/gateway` is the same server holding the same seqs; re-downloading everything
         * because a path changed is a cost with nothing behind it.
         */
        public fun of(endpoint: String, appId: String, userId: String): ImCursorScope =
            ImCursorScope(
                host = endpoint.substringAfter("://").substringBefore('/').substringBefore('?'),
                appId = appId,
                userId = userId,
            )
    }
}

/**
 * What survives a process restart.
 *
 * @property convSeqs per conversation, the highest seq the **application has durably stored**.
 *   This is what `conn.sync` is told, and it is deliberately not "the highest seq we received":
 *   a message that reached a callback and not a database is exactly what a crash loses.
 * @property conversationCursor the highest `ConversationView.updatedAt` a **completed**
 *   `conn.sync` run consumed. Advancing it after a partial run permanently hides every
 *   conversation on the pages that were never read.
 * @property scope which [ImCursorScope] these belong to, as [ImCursorScope.key]. The SDK stamps it
 *   on every write and refuses a snapshot belonging to a different account, which is what stops
 *   account switching on a shared handset from handing one user another's cursors.
 *
 *   **The identity lives in the payload rather than in the store's signature on purpose.** A store
 *   is a few lines an integrator writes; a store that keys itself per account is a few lines an
 *   integrator *remembers* to write. Carrying the identity here means the SDK can refuse cursors
 *   belonging to somebody else even when the store did nothing to keep two accounts apart — an app
 *   that does nothing special cannot get it wrong. Null on a snapshot the SDK has never written.
 */
@Serializable
public data class ImCursorSnapshot(
    public val convSeqs: Map<String, Long> = emptyMap(),
    public val conversationCursor: Long = 0,
    public val scope: String? = null,
)

/**
 * Where the cold-start cursors live.
 *
 * **This is a required constructor argument of [ImClient], and there is no default.** That is a
 * deliberate compile error rather than a convenience.
 *
 * The bug it exists to prevent: a client that starts with no cursors reports nothing in
 * `convSeqs`, the server therefore reports no gaps (`ConnController.Sync` only produces a gap for
 * a conversation the client reported a non-zero seq for), and the client then adopts the server's
 * current `maxSeq`. Every message that arrived while the app was closed is now behind the cursor,
 * will never be requested and will never arrive — with no error, no log line and no later event
 * that corrects it. Four of the five SDKs shipped that behaviour.
 *
 * Defaulting to no persistence is what produced it. Defaulting to *some* persistence would mean
 * the SDK guessing where a JVM application may write, and a wrong guess about that is worse than
 * a compile error. So the choice is yours and it is explicit: [file] for a real one, [inMemory]
 * for none.
 *
 * **Lifetime: this store and your own message store have exactly one lifetime.** Whatever destroys
 * one destroys the other, cursors first. Clearing messages while keeping cursors shows an empty
 * conversation that will never refill; clearing cursors while keeping messages costs a full
 * re-download and duplicate delivery. A logout that keeps messages for a fast re-login keeps the
 * cursors too.
 *
 * 本存储与你自己的消息库共用同一个生命周期，销毁时先游标后消息。只清消息会留下一个
 * 永远填不回来的空会话；只清游标会导致全量重下与重复投递。
 *
 * Implementations may be called from any thread but never concurrently with themselves; the SDK
 * serialises its own writes. [save] runs on the dispatcher passed to [ImClient] as
 * `cursorStoreContext`, which defaults to `Dispatchers.IO`, so blocking here is expected.
 */
public interface ImCursorStore {

    /** Called once, before the first connect. Throwing is handled — see [ImClient.cursorStoreState]. */
    public fun load(): ImCursorSnapshot

    /** Called by the SDK. May be coalesced; see [ImOptions.cursorFlushInterval]. */
    public fun save(snapshot: ImCursorSnapshot)

    public companion object {
        /**
         * No persistence at all. Every cold start adopts the server's position and silently drops
         * whatever arrived while the process was down.
         *
         * Legitimate for a bot, a test, or a client whose message store is itself in memory.
         * Not legitimate for a chat app — pass [file] and give it a directory.
         */
        public fun inMemory(): ImCursorStore = InMemoryCursorStore()

        /**
         * One JSON file, written through a temp file and renamed.
         *
         * The host app supplies the path because this artifact has no Android dependency and
         * therefore no `Context`: on Android pass `File(context.filesDir, "im-cursors.json")`, on
         * a desktop or service host whatever directory you already own.
         *
         * `java.io.File` rather than `java.nio.file.Path`, deliberately. `Path` and `Files` are
         * API 26 on Android and this SDK supports API 21; a `Path` in this signature would compile
         * everywhere and then throw `NoClassDefFoundError` on a five-year-old handset unless the
         * app happened to have core library desugaring switched on.
         *
         * The file is per account. If you point two accounts at the same path the SDK notices —
         * the snapshot carries its scope key — and starts clean rather than adopting the other
         * user's cursors, but you lose the first account's cursors when the second one writes.
         * Put the user id in the filename.
         */
        public fun file(path: File): ImCursorStore = FileCursorStore(path)

        /** [file] for callers holding a string path. */
        public fun file(path: String): ImCursorStore = FileCursorStore(File(path))
    }
}

/** What happened when [ImCursorStore.load] was called. Observable so an app can assert on it. */
public sealed interface ImCursorStoreState {

    /** Before the first connect. */
    public data object NotLoaded : ImCursorStoreState

    /**
     * The snapshot was read. [conversations] is how many cursors came back — **zero on a genuine
     * fresh install, and also zero when a store silently loses its data**, which is why
     * [persistent] is here too.
     */
    public data class Loaded(
        public val conversations: Int,
        public val conversationCursor: Long,
        public val persistent: Boolean,
    ) : ImCursorStoreState

    /**
     * [ImCursorStore.load] threw, or returned something the SDK could not use.
     *
     * The SDK does **not** treat this as a fresh install: it refuses to adopt any conversation and
     * refuses to advance any cursor for the rest of the session. A failed load and a fresh install
     * are indistinguishable to the adoption branch, and adopting on a failed load destroys history
     * that is sitting intact in your own database.
     *
     * You decide what to do: re-derive the cursors from your own message store and replay them
     * through [ImClient.commit] (the correct fix, and the reason `commit` is monotonic), or accept
     * a re-download by constructing a new client with a working store.
     */
    public data class Failed(public val cause: Throwable) : ImCursorStoreState
}

/**
 * [ImCursorStore.inMemory].
 *
 * Recognised by [ImClient] by type, and that is the whole mechanism: "is this store persistent" is
 * answered by the SDK from the store's identity, never declared by the store. The interface used to
 * carry an `isPersistent` flag, which let *any* store — including one an integrator wrote over
 * SQLite — announce itself volatile; and the only thing that answer drives is the warning about
 * losing messages. A store that can lie about that is a store that can silence the one line
 * standing between a misconfigured client and a support ticket about missing history.
 *
 * 「是否持久化」由 SDK 按存储的身份判断，不由存储自己声明。
 */
internal class InMemoryCursorStore : ImCursorStore {
    private val held = AtomicReference(ImCursorSnapshot())

    override fun load(): ImCursorSnapshot = held.get()

    override fun save(snapshot: ImCursorSnapshot) {
        held.set(snapshot)
    }
}

/**
 * [ImCursorStore.file].
 *
 * Written through a sibling temp file and then moved, because the failure this whole mechanism
 * guards against is a *partial* cursor file: a half-written JSON document fails to parse, which
 * the SDK correctly refuses to treat as a fresh install, which strands the session with no cursor
 * advancement until the app is restarted. A rename is the cheapest way to make the file either the
 * old one or the new one and never something in between.
 */
internal class FileCursorStore(private val file: File) : ImCursorStore {

    override fun load(): ImCursorSnapshot {
        if (!file.exists()) return ImCursorSnapshot()
        val text = file.readText(Charsets.UTF_8)
        if (text.isBlank()) return ImCursorSnapshot()
        return ImJson.decodeFromString(ImCursorSnapshot.serializer(), text)
    }

    override fun save(snapshot: ImCursorSnapshot) {
        file.parentFile?.mkdirs()
        val temp = File(file.parentFile, "${file.name}.tmp")
        temp.writeText(ImJson.encodeToString(ImCursorSnapshot.serializer(), snapshot), Charsets.UTF_8)

        // On every filesystem Android uses, renaming over an existing file is atomic and is what
        // makes the file either the old snapshot or the new one. The fallbacks are for Windows and
        // the odd FUSE volume, which refuse a rename onto an existing name.
        if (temp.renameTo(file)) return
        if (file.delete() && temp.renameTo(file)) return

        try {
            temp.copyTo(file, overwrite = true)
        } finally {
            temp.delete()
        }
    }
}

File: unity/Runtime/ImCursorStore.cs
using System;
using System.Collections.Generic;
using System.Globalization;
using System.IO;
using System.Text;
using Cyaim.Im.Json;
using UnityEngine;

namespace Cyaim.Im
{
    /// <summary>
    /// Where the SDK keeps the two numbers that decide what a cold start downloads.
    /// </summary>
    /// <remarks>
    /// <para>
    /// This interface exists because of a specific data-loss bug, described in full in
    /// <c>sdk/CONTRACT.md</c> §5.1. A client that starts with no cursors reports nothing in
    /// <c>conn.sync</c>'s <c>convSeqs</c>; the server can only compute a gap for a conversation the
    /// client reported, so it reports none; the client then takes the server's <c>maxSeq</c> as its
    /// position, and every message that arrived while the app was closed is now behind the cursor.
    /// It is never requested, never arrives, and nothing anywhere says so.
    /// </para>
    /// <para>
    /// Which is why the store is a <b>required</b> option (<see cref="ImClientOptions.CursorStore"/>)
    /// rather than something with a default. There is no safe default: a wrong guess about where a
    /// game may write is worse than a loud error at construction, and defaulting to "no persistence"
    /// is exactly what produced the bug. An integrator who genuinely does not want persistence
    /// passes <see cref="ImCursorStore.InMemory"/>, which says so once in the log.
    /// </para>
    /// <para>
    /// <b>Lifetime.</b> This store and the game's own message store have exactly one lifetime.
    /// Whatever destroys one destroys the other, cursors first. Clearing messages but keeping
    /// cursors leaves a conversation the app will never refill; clearing cursors but keeping
    /// messages costs a full re-download and duplicate delivery.
    /// </para>
    /// <para>
    /// 冷启动丢消息的根因见 CONTRACT §5.1：客户端不报 convSeqs，服务端就算不出缺口，
    /// 客户端再把服务端的 maxSeq 当作自己的位置——关机期间到达的消息就此永久落在游标后面。
    /// </para>
    /// </remarks>
    public interface IImCursorStore
    {
        /// <summary>
        /// Reads the stored snapshot. Called once, before the first connect.
        /// </summary>
        /// <remarks>
        /// <para>
        /// Throwing is a legitimate answer and is <i>not</i> treated as "fresh install" — see
        /// <see cref="ImClient.CursorStoreFailed"/>. A failed load and a fresh install look
        /// identical to the adoption branch, and adopting on a failed load destroys history that is
        /// sitting intact in the game's own database.
        /// </para>
        /// <para>
        /// There is no <c>scope</c> parameter. The interface is exactly <c>Load</c> / <c>Save</c> —
        /// the shape <c>sdk/CONTRACT.md</c> §5.3 specifies, and the same two members the other
        /// four SDKs expose. The account identity travels inside
        /// <see cref="ImCursorSnapshot.Scope"/> instead, which the SDK stamps on every write and
        /// checks on every read: a store handed the scope as an argument can still ignore it, and a
        /// store that never sees it cannot defeat the check at all.
        /// </para>
        /// </remarks>
        ImCursorSnapshot Load();

        /// <summary>
        /// Persists the snapshot. The SDK may call this often; see
        /// <see cref="ImClientOptions.CursorFlushInterval"/> for what is coalesced and what is not.
        /// </summary>
        void Save(ImCursorSnapshot snapshot);
    }

    /// <summary>
    /// Which account's cursors are being read or written: endpoint host, app id, user id.
    /// </summary>
    /// <remarks>
    /// <para>
    /// The device id is deliberately <i>not</i> part of the key — the store is already local to the
    /// device. The user id is, and that is the point: without it, two accounts sharing a handset
    /// read each other's cursors, and the second one to sign in silently skips whatever the first
    /// one had already consumed.
    /// </para>
    /// <para>
    /// The SDK stamps <see cref="Key"/> into every snapshot it writes and refuses one carrying
    /// somebody else's, so the guarantee holds even for a store that knows nothing about scoping.
    /// <see cref="StorageKey"/> is for the built-in stores that pick their own file name and want
    /// one file per account — which is what makes signing back *in* to an earlier account find
    /// its cursors instead of re-downloading.
    /// </para>
    /// </remarks>
    public readonly struct ImCursorScope
    {
        /// <summary>Gateway host, without scheme or port.</summary>
        public string Host { get; }

        /// <summary>Tenant id.</summary>
        public string AppId { get; }

        /// <summary>The signed-in user. Empty when the host app never set
        /// <see cref="ImConnectionOptions.UserId"/>, which the SDK warns about once.</summary>
        public string UserId { get; }

        /// <inheritdoc cref="ImCursorScope"/>
        public ImCursorScope(string host, string appId, string userId)
        {
            Host = host ?? string.Empty;
            AppId = appId ?? string.Empty;
            UserId = userId ?? string.Empty;
        }

        /// <summary>
        /// Builds a scope from an endpoint URL, taking only its host. The scheme and port are left
        /// out on purpose: moving a deployment from <c>ws://</c> to <c>wss://</c>, or off a
        /// development port, must not orphan the cursors already on the device.
        /// </summary>
        public static ImCursorScope Of(string endpoint, string appId, string userId)
        {
            return new ImCursorScope(HostOf(endpoint), appId, userId);
        }

        /// <summary>Former name of <see cref="Of"/>. The other four SDKs spell it <c>of</c>.</summary>
        [Obsolete("Renamed to ImCursorScope.Of, matching the other four SDKs. Removed in 2.0.")]
        public static ImCursorScope For(string endpoint, string appId, string userId)
        {
            return Of(endpoint, appId, userId);
        }

        /// <summary>
        /// The identity stamped into <see cref="ImCursorSnapshot.Scope"/> and compared on load:
        /// <c>host|appId|userId</c>. Spelled the same way in all five SDKs, so a snapshot written by
        /// one is legible to the others.
        /// </summary>
        public string Key
        {
            get
            {
                return Host + "|" + AppId + "|" + (string.IsNullOrEmpty(UserId) ? "*" : UserId);
            }
        }

        /// <summary>
        /// A stable, filesystem- and PlayerPrefs-safe rendering of the same identity. Same three
        /// inputs always produce the same string, on every platform.
        /// </summary>
        public string StorageKey
        {
            get
            {
                var key = new StringBuilder(96);
                Append(key, Host);
                key.Append('_');
                Append(key, AppId);
                key.Append('_');
                Append(key, string.IsNullOrEmpty(UserId) ? "anonymous" : UserId);
                return key.ToString();
            }
        }

        /// <inheritdoc/>
        public override string ToString()
        {
            return Host + "/" + AppId + "/" + (string.IsNullOrEmpty(UserId) ? "(no userId)" : UserId);
        }

        private static string HostOf(string endpoint)
        {
            if (string.IsNullOrEmpty(endpoint))
            {
                return string.Empty;
            }

            Uri parsed;
            if (Uri.TryCreate(endpoint, UriKind.Absolute, out parsed))
            {
                return parsed.Host;
            }

            return endpoint;
        }

        /// <summary>
        /// Copies a segment, replacing anything that is not plainly safe. Ids are tenant-chosen and
        /// have been seen to contain slashes and colons; one of those in a file name is a write that
        /// lands somewhere else or not at all.
        /// </summary>
        private static void Append(StringBuilder target, string value)
        {
            if (string.IsNullOrEmpty(value))
            {
                target.Append('-');
                return;
            }

            for (int i = 0; i < value.Length; i++)
            {
                var c = value[i];
                var safe = (c >= 'a' && c <= 'z') || (c >= 'A' && c <= 'Z') || (c >= '0' && c <= '9') ||
                           c == '-' || c == '.';
                target.Append(safe ? c : '_');
            }
        }
    }

    /// <summary>
    /// What the SDK persists: how far each conversation is durably stored, and how far the
    /// conversation list itself has been consumed.
    /// </summary>
    public sealed class ImCursorSnapshot
    {
        /// <summary>
        /// conversationId to <c>committedSeq</c> — the highest seq the <i>application</i> has said
        /// it has durably stored, never the highest the SDK has delivered. This is the map reported
        /// as <c>convSeqs</c>, and reporting the delivered one instead is the bug in §5.1.
        /// </summary>
        public Dictionary<string, long> ConvSeqs { get; private set; }

        /// <summary>
        /// Highest <c>ConversationView.updatedAt</c> from a <i>completed</i> <c>conn.sync</c> run.
        /// </summary>
        /// <remarks>
        /// The list is sorted by <c>updatedAt</c> descending, so page one holds the newest value.
        /// Advancing this before the run finishes pushes the cursor past every conversation on the
        /// pages that were never read, and the server will never return them again.
        /// </remarks>
        public long ConversationCursor { get; set; }

        /// <summary>
        /// Which <see cref="ImCursorScope"/> these cursors belong to, as
        /// <see cref="ImCursorScope.Key"/>. Stamped by the SDK on every write.
        /// </summary>
        /// <remarks>
        /// <para>
        /// <b>This field is the account-switch guarantee, and it lives in the payload rather than in
        /// the store's signature on purpose.</b> A store is a few lines an integrator writes; a
        /// store that keys itself per account is a few lines an integrator <i>remembers</i> to
        /// write. Carrying the identity here means the SDK can refuse cursors belonging to somebody
        /// else even when the store did nothing to keep two accounts apart — so a game that does
        /// nothing special cannot get it wrong.
        /// </para>
        /// <para>
        /// Null or empty on a snapshot the SDK has never written. That is accepted, and the next
        /// save stamps it.
        /// </para>
        /// </remarks>
        public string Scope { get; set; }

        /// <inheritdoc cref="ImCursorSnapshot"/>
        public ImCursorSnapshot()
        {
            ConvSeqs = new Dictionary<string, long>(StringComparer.Ordinal);
        }

        /// <summary>An independent copy, so a store can serialise off the SDK's live map.</summary>
        public ImCursorSnapshot Clone()
        {
            var copy = new ImCursorSnapshot { ConversationCursor = ConversationCursor, Scope = Scope };
            foreach (var entry in ConvSeqs)
            {
                copy.ConvSeqs[entry.Key] = entry.Value;
            }

            return copy;
        }

        /// <summary>Serialises to the on-disk shape used by the built-in stores.</summary>
        public string ToJson()
        {
            var seqs = JsonValue.NewObject();
            foreach (var entry in ConvSeqs)
            {
                seqs.Set(entry.Key, entry.Value);
            }

            var root = JsonValue.NewObject()
                .Set("version", 1L)
                .Set("conversationCursor", ConversationCursor)
                .Set("convSeqs", seqs);

            if (!string.IsNullOrEmpty(Scope))
            {
                root.Set("scope", Scope);
            }

            return root.ToJson();
        }

        /// <summary>
        /// Reads the on-disk shape. Throws <see cref="ImJsonException"/> on text that is not JSON —
        /// deliberately, because a truncated file must reach
        /// <see cref="ImClient.CursorStoreFailed"/> rather than be mistaken for a fresh install.
        /// </summary>
        public static ImCursorSnapshot FromJson(string text)
        {
            var root = JsonValue.Parse(text);
            if (!root.IsObject)
            {
                throw new ImJsonException("a cursor snapshot must be a JSON object", 0);
            }

            var snapshot = new ImCursorSnapshot
            {
                ConversationCursor = root["conversationCursor"].AsLong(),
                Scope = root["scope"].AsString(null),
            };

            foreach (var member in root["convSeqs"].Members)
            {
                // Zero entries are kept, not dropped. A conversation that existed with no messages
                // yet is a conversation this device has already been told about, and forgetting it
                // would make the next start treat it as first sight and adopt whatever maxSeq it
                // has grown to — skipping every message it gained in between. Zero is filtered only
                // where it goes onto the wire, because there the server reads absent and 0 alike.
                snapshot.ConvSeqs[member.Key] = member.Value.AsLong();
            }

            return snapshot;
        }
    }

    /// <summary>
    /// The stores that ship with the SDK. Each one is an explicit choice; there is no default.
    /// </summary>
    public static class ImCursorStore
    {
        /// <summary>Folder under <c>Application.persistentDataPath</c> used when none is named.</summary>
        public const string DefaultSubdirectory = "cyaim-im";

        /// <summary>
        /// Keeps cursors under <c>Application.persistentDataPath</c>, one file per account. The
        /// right answer for almost every game.
        /// </summary>
        /// <remarks>
        /// On WebGL this delegates to <see cref="PlayerPrefs"/>: a browser build's
        /// <c>persistentDataPath</c> is an IndexedDB image that Unity flushes on its own schedule,
        /// and a cursor written into a flush that never happened is the §5.5 failure exactly.
        /// <c>PlayerPrefs</c> is synchronous there and is what the device id already uses.
        /// </remarks>
        /// <param name="scope">
        /// Whose cursors these are. Required, because this factory picks the file name and keys it
        /// per account: signing back in as an earlier account then finds its cursors intact rather
        /// than re-baselining to the server's position. Build it with
        /// <see cref="ImCursorScope.Of"/> from the same endpoint, app id and user id you pass to
        /// <see cref="ImClientOptions"/>.
        /// </param>
        /// <param name="subdirectory">Folder name under the persistent data path.</param>
        public static IImCursorStore PersistentDataPath(
            ImCursorScope scope,
            string subdirectory = DefaultSubdirectory)
        {
            if (Application.platform == RuntimePlatform.WebGLPlayer)
            {
                return PlayerPrefs(scope);
            }

            var root = Path.Combine(Application.persistentDataPath, string.IsNullOrEmpty(subdirectory)
                ? DefaultSubdirectory
                : subdirectory);

            return new ImFileCursorStore(Path.Combine(root, scope.StorageKey + ".cursors.json"));
        }

        /// <summary>
        /// Keeps cursors in one file at a path you choose. For a headless build, a dedicated server,
        /// or an editor tool that has nowhere else to put them.
        /// </summary>
        /// <remarks>
        /// The path is used verbatim, so a host that runs two accounts through one process must give
        /// each its own path — unlike <see cref="PersistentDataPath"/>, which derives the file name
        /// from the scope and cannot collide. Point two accounts at one path and the SDK notices,
        /// because the snapshot carries its scope, and starts clean rather than handing the second
        /// user the first one's cursors; the first account's cursors are gone once the second one
        /// writes. <c>ImCursorScope.Of(…).StorageKey</c> in the file name is the whole fix.
        /// </remarks>
        public static IImCursorStore File(string path)
        {
            if (string.IsNullOrEmpty(path))
            {
                throw new ArgumentException("path is required", "path");
            }

            return new ImFileCursorStore(path);
        }

        /// <summary>Keeps cursors in <c>UnityEngine.PlayerPrefs</c>, one entry per account.</summary>
        /// <remarks>
        /// Small and synchronous, and the only thing that reliably survives on WebGL.
        /// <c>PlayerPrefs</c> is not a database: a snapshot for a user with thousands of
        /// conversations belongs in a file.
        /// </remarks>
        public static IImCursorStore PlayerPrefs(
            ImCursorScope scope,
            string keyPrefix = "cyaim.im.cursors.")
        {
            return new ImPlayerPrefsCursorStore((keyPrefix ?? string.Empty) + scope.StorageKey);
        }

        /// <summary>
        /// Keeps cursors for the life of the process and no longer. An explicit opt-out of
        /// persistence, and it logs one warning naming <c>sdk/CONTRACT.md</c> §5.3 to say so.
        /// </summary>
        /// <remarks>
        /// Legitimate for a bot, a load test, or a kiosk build that has no local message store to
        /// keep in step. In a player-facing app it means every cold start re-adopts the server's
        /// position, which is the data loss this whole mechanism exists to prevent.
        /// </remarks>
        public static IImCursorStore InMemory()
        {
            return new ImMemoryCursorStore();
        }
    }

    /// <summary>Snapshots held in a field. See <see cref="ImCursorStore.InMemory"/>.</summary>
    internal sealed class ImMemoryCursorStore : IImCursorStore
    {
        private ImCursorSnapshot _held;

        public ImCursorSnapshot Load()
        {
            // No warning here. <see cref="ImClient"/> raises the one §5.3 asks for, because whether
            // the game is told it is running without persistence is not a store's decision to make:
            // a store that could answer that question could also answer it wrongly, and three of
            // the five SDKs let any custom store do exactly that.
            return _held != null ? _held.Clone() : new ImCursorSnapshot();
        }

        public void Save(ImCursorSnapshot snapshot)
        {
            _held = snapshot.Clone();
        }
    }

    /// <summary>One JSON file per account. See <see cref="ImCursorStore.PersistentDataPath"/>.</summary>
    internal sealed class ImFileCursorStore : IImCursorStore
    {
        private readonly string _path;

        internal ImFileCursorStore(string path)
        {
            _path = path;
        }

        public ImCursorSnapshot Load()
        {
            var path = _path;
            if (!System.IO.File.Exists(path))
            {
                // Genuinely absent, which is a fresh install and the one case where adoption is
                // correct. Unreadable is a different answer and is allowed to throw.
                return new ImCursorSnapshot();
            }

            return ImCursorSnapshot.FromJson(System.IO.File.ReadAllText(path, Encoding.UTF8));
        }

        public void Save(ImCursorSnapshot snapshot)
        {
            var path = _path;
            var folder = Path.GetDirectoryName(path);
            if (!string.IsNullOrEmpty(folder) && !Directory.Exists(folder))
            {
                Directory.CreateDirectory(folder);
            }

            // Write beside the target and swap. A half-written cursor file is worse than an old
            // one: the old one costs duplicate delivery, the truncated one costs the whole
            // snapshot, and a game is killed mid-write by the OS more often than anyone expects.
            var temporary = path + ".tmp";
            System.IO.File.WriteAllText(temporary, snapshot.ToJson(), Encoding.UTF8);

            try
            {
                if (System.IO.File.Exists(path))
                {
                    System.IO.File.Replace(temporary, path, null);
                }
                else
                {
                    System.IO.File.Move(temporary, path);
                }
            }
            catch (Exception error)
            {
                // Some platforms and some filesystems do not implement Replace. Falling back is
                // strictly worse — there is a window where neither file is the snapshot — but it is
                // better than never writing at all.
                ImLog.Warn("atomic cursor write is unavailable here; falling back to delete-and-move", error);
                System.IO.File.Delete(path);
                System.IO.File.Move(temporary, path);
            }
        }
    }

    /// <summary>One PlayerPrefs entry per account. See <see cref="ImCursorStore.PlayerPrefs"/>.</summary>
    internal sealed class ImPlayerPrefsCursorStore : IImCursorStore
    {
        private readonly string _prefix;

        internal ImPlayerPrefsCursorStore(string prefix)
        {
            _prefix = prefix;
        }

        public ImCursorSnapshot Load()
        {
            var text = UnityEngine.PlayerPrefs.GetString(_prefix, null);
            return string.IsNullOrEmpty(text) ? new ImCursorSnapshot() : ImCursorSnapshot.FromJson(text);
        }

        public void Save(ImCursorSnapshot snapshot)
        {
            UnityEngine.PlayerPrefs.SetString(_prefix, snapshot.ToJson());

            // Save() rather than leaving it to the player loop: the writes that matter most are the
            // ones taken as the app is being suspended, and an unflushed PlayerPrefs is discarded
            // when the OS kills the process.
            UnityEngine.PlayerPrefs.Save();
        }
    }

    /// <summary>
    /// Raised when the cursor store could not be read. Carries what the store threw.
    /// </summary>
    /// <remarks>
    /// The SDK's response to this is deliberately conservative and is described in
    /// <c>sdk/CONTRACT.md</c> §5.8: it refuses to adopt any conversation and refuses to advance any
    /// cursor for the rest of the session. Delivery keeps working; nothing is persisted. The app
    /// decides whether to re-derive cursors from its own store — the correct fix, and the reason
    /// <see cref="ImClient.Commit"/> is monotonic — or to accept a re-download next launch.
    /// </remarks>
    public sealed class ImCursorStoreException : Exception
    {
        /// <summary>The account whose snapshot could not be read.</summary>
        public ImCursorScope Scope { get; private set; }

        /// <inheritdoc cref="ImCursorStoreException"/>
        public ImCursorStoreException(ImCursorScope scope, Exception inner)
            : base("the cursor store could not be read for " + scope +
                   "; no conversation will be adopted and no cursor will be persisted this session " +
                   "(sdk/CONTRACT.md §5.8)", inner)
        {
            Scope = scope;
        }
    }

    /// <summary>Numbers the SDK reports about itself. Quote <see cref="ContractVersion"/> in a ticket.</summary>
    public static class ImSdk
    {
        /// <summary>
        /// Version of <c>sdk/CONTRACT.md</c> this build implements. Sourced from
        /// <c>sdk/endpoint-inventory.json</c> and identical across all five client SDKs, so it
        /// answers "which endpoints do you have" without anyone having to ask which platform.
        /// </summary>
        public const string ContractVersion = "1.0";

        /// <summary>
        /// This package's version, sent as the <c>cv</c> handshake parameter by default. Kept equal
        /// to <c>package.json</c> by <c>VersionGuardTests</c>, which also checks that all five SDKs
        /// carry the same number (<c>sdk/CONTRACT.md</c> §9.1).
        /// </summary>
        public const string PackageVersion = "0.9.0";
    }
}
